File: code/python/ch04/solutions/warmup_05_sol.py
"""
Solution for Warmup 5: Double List
============================================
Chapter 4: Functions

This is the reference solution. Try to solve the problem yourself before
looking at this!

APPROACH
--------
Loop through the list by index and multiply each element by 2 in place.
Return the same list object (not a new list).

TIME COMPLEXITY:  O(n) where n = len(nums)
SPACE COMPLEXITY: O(1) — modifies in place
"""


def solve(nums: list[int]) -> list[int]:
    """Double every element in nums in place and return the list."""
    for i in range(len(nums)):
        nums[i] *= 2
    return nums


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    line = input().strip()
    if line:
        nums = list(map(int, line.split()))
    else:
        nums = []
    result = solve(nums)
    print(" ".join(map(str, result)))

File: code/java/ch07/solutions/Warmup05Sol.java
package ch07.solutions;

import java.util.*;

/**
 * Solution for Warmup 05: Armstrong Number
 * =========================================
 * Chapter 7: Number Wizardry
 *
 * This is the reference solution. Try to solve the problem yourself before
 * looking at this!
 *
 * APPROACH
 * --------
 * Count the digits (k), then sum each digit raised to the power k.
 * Compare the sum with the original number.
 *
 * TIME COMPLEXITY:  O(d) where d = number of digits
 * SPACE COMPLEXITY: O(1)
 */
public class Warmup05Sol {

    public static boolean solve(long n) {
        if (n < 0) return false;
        int numDigits = String.valueOf(n).length();
        long temp = n;
        long sum = 0;
        while (temp > 0) {
            long d = temp % 10;
            sum += (long) Math.pow(d, numDigits);
            temp /= 10;
        }
        return sum == n;
    }

    // ── Do not change anything below this line ──────────────────────
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        long n = Long.parseLong(sc.nextLine().trim());
        System.out.println(solve(n));
        sc.close();
    }
}
